edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/InvokeMethodHandler.cs

```diff
@@ -0,0 +1,157 @@
+                        {
+                            taskCompletion.TrySetResult(new DirectMethodResponse(new EdgeHubTimeoutException($"Client {methodRequest.Id} not found"), HttpStatusCode.NotFound));
+                        }
+
+                        return await taskCompletion.Task;
+                    });
+        }
+
+        public Task ProcessInvokeMethodSubscription(string id)
+        {
+            this.ProcessInvokeMethodSubscriptionInternal(id);
+            return Task.CompletedTask;
+        }
+
+        async void ProcessInvokeMethodSubscriptionInternal(string id)
+        {
+            try
+            {
+                Events.ProcessingInvokeMethodQueue(id);
+                // Temporary hack to wait for the subscription call to complete. Without this,
+                // the EdgeHub will invoke the pending method request "too soon", before the layers
+                // in between have been set up correctly. To fix this, changes are needed in ProtocolGateway,
+                // Client SDK, and need to figure out a way to raise events for AMQP Links. 
+                await Task.Delay(TimeSpan.FromSeconds(3));
+                await this.ProcessInvokeMethodsForClient(id);
+            }
+            catch (Exception e)
+            {
+                Events.ErrorProcessingInvokeMethodRequests(e, id);
+            }
+        }
+
+        Option<IDeviceProxy> GetDeviceProxyWithSubscription(string id)
+        {
+            Option<IDeviceProxy> deviceProxy = this.connectionManager.GetDeviceConnection(id);
+            if (!deviceProxy.HasValue)
+            {
+                Events.NoDeviceProxyForMethodInvocation(id);
+            }
+            else if (!this.connectionManager.GetSubscriptions(id)
+                .Filter(s => s.TryGetValue(DeviceSubscription.Methods, out bool isActive) && isActive)
+                .HasValue)
+            {
+                Events.NoSubscriptionForMethodInvocation(id);
+            }
+            else
+            {
+                return deviceProxy;
+            }
+            return Option.None<IDeviceProxy>();
+        }
+
+        async Task ProcessInvokeMethodsForClient(string id)
+        {
+            Option<IDeviceProxy> deviceProxy = this.GetDeviceProxyWithSubscription(id);
+            await deviceProxy.ForEachAsync(
+                async dp =>
+                {
+                    ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>> clientQueue = this.GetClientQueue(id);
+                    foreach (KeyValuePair<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>> invokeMethodRequest in clientQueue)
+                    {
+                        if (clientQueue.TryRemove(invokeMethodRequest.Key, out TaskCompletionSource<DirectMethodResponse> taskCompletionSource))
+                        {
+                            DirectMethodResponse response = await dp.InvokeMethodAsync(invokeMethodRequest.Key);
```

Comment:
Should the argument here be `invokeMethodRequest` instead of `invokeMethodRequest.Key`?